File: src/main/nodejs/serve.js
import express from 'express';
import fs from 'fs';
import jose from 'node-jose';  // Changed to default import
import chalk from 'chalk';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';

// Parse command line arguments
const argv = yargs(hideBin(process.argv))
    .option('read-key', {
        describe: 'Path to private PEM key file',
        type: 'string',
        demandOption: true
    })
    .parse();

const keyPath = argv['read-key'];

// Verify key file exists
if (!fs.existsSync(keyPath)) {
    console.error(`Private PEM key not found in ${keyPath}`);
    console.error('You could generate one with: openssl genpkey -algorithm RSA -pkeyopt rsa_keygen_bits:2048 -out private_key.pem');
    process.exit(1);
}

const app = express();

// Add logging middleware
app.use((req, res, next) => {
    const clientIp = req.ip || req.socket.remoteAddress;
    console.log(chalk.blue(`[${new Date().toISOString()}]`) + ` ` + chalk.magenta(`${clientIp}`) + ` => ` + chalk.green(`${req.method}`) +` ${req.path}`);
    next();
});

// Read and convert PEM to JWK
async function readPemKey() {
    const pemData = fs.readFileSync(keyPath, 'utf8');
    const keystore = await jose.JWK.createKeyStore();
    return await keystore.add(Buffer.from(pemData), 'pem');
}

// Get public and private keys
function getKeysFrom(key) {
    return {
        ...key.toJSON(false),  // Include key parameters
        alg: 'RS256',
        use: 'sig'
    };
}

// Define JWKS endpoint
app.get('/.well-known/jwks.json', async (req, res) => {
    const key = await readPemKey();
    const publicKey = getKeysFrom(key);
    res.json({ keys: [publicKey] });
});

// Start server
const PORT = 8085;
app.listen(PORT, '0.0.0.0', () => {
    console.log(chalk.red.bold('This server is for demo purposes only. Do not use as is in production'));
    console.log(`Server running on port ${PORT}`);
});

File: src/main/python/ubench/jwks/serve.py
import getopt
import sys
import string
import os
import random
import rich
from rich.console import Console
from flask import Flask, jsonify
from jwcrypto import jwk

console = Console()

def read_pem_key():
    # check if key file exists        
    with open(key_path) as f:
        pem_data = f.read()
    f.closed

    pem_data_encode = pem_data.encode("utf-8")
    return jwk.JWK.from_pem(pem_data_encode)


def get_keys_from(key):
    private_key = key.export_private()
    public_key = key.export_public(as_dict=True)
    public_key['alg'] = 'RS256'
    public_key['use'] = 'sig'
    return (public_key, private_key)



opts, args = getopt.getopt(sys.argv[1:], "", ["read-key="])

key_function = None
key_path:str = None

for opt, arg in opts:
    if opt in ("--read-key"):
        key_function = read_pem_key
        key_path = arg
        if not key_path:
            raise Exception("Path to your private key not specified: Use --key-path=<key-path>")
        
        if not os.path.isfile(key_path):
            raise Exception(f"Private pem key not found in {key_path}.\nYou could generate one with `openssl genpkey -algorithm RSA -pkeyopt rsa_keygen_bits:2048 -out private_key.pem`")

if key_function is None:
    raise Exception("key function not specified: Use --read-key=<path-to-private-pem-key>")


app = Flask(__name__)

# Define a route for the JWKS endpoint
@app.route('/.well-known/jwks.json')
def jwks():
    public_key, _ = get_keys_from(key_function())
    return jsonify({"keys": [public_key]})

console.print("This server is for demo purposes only. Do not use as is in production", style="bold red")

# Run the Flask application
if __name__ == '__main__':
    app.run(host='0.0.0.0', port=8085)
